modules/contracts/apps/challenge/src/handlers/execute.rs

```diff
@@ -0,0 +1,568 @@
+
+        return Ok(Response::new()
+            .add_attribute("action", "tally_vote")
+            .add_attribute(
+                "message",
+                "All votes were positive. ChallengeStatus has been set to OverAndCompleted.",
+            ));
+    }
+}
+
+fn charge_penalty(
+    deps: DepsMut,
+    info: MessageInfo,
+    app: &ChallengeApp,
+    challenge_id: u64,
+) -> AppResult {
+    deps.api.addr_validate(info.sender.as_ref())?;
+
+    let challenge = CHALLENGE_LIST.load(deps.storage, challenge_id.clone())?;
+    if challenge.status != ChallengeStatus::OverAndFailed {
+        return Err(AppError::WrongChallengeStatus {});
+    }
+    let challenge = CHALLENGE_LIST.load(deps.storage, challenge_id.clone())?;
+    let friends = CHALLENGE_FRIENDS.load(deps.storage, challenge_id.clone())?;
+
+    let bank = app.bank(deps.as_ref());
+    let executor = app.executor(deps.as_ref());
+
+    match challenge.collateral {
+        Penalty::FixedAmount { asset } => {
+            let num_friends = friends.len() as u128;
+            if num_friends == 0 {
+                return Err(AppError::Std(StdError::generic_err(
+                    "No friends found for the challenge.",
+                )));
+            }
+
+            // Calculate each friend's share
+            let amount_per_friend = asset.amount.u128() / num_friends;
+            let reaminder = asset.amount.u128() % num_friends;
+            let asset_per_friend = OfferAsset {
+                name: asset.name,
+                amount: Uint128::from(amount_per_friend),
+            };
+
+            // Create a transfer action for each friend
+            let transfer_actions: Result<Vec<_>, _> = friends
+                .into_iter()
+                .map(|friend| bank.transfer(vec![asset_per_friend.clone()], &friend.address))
+                .collect();
+
+            let transfer_msgs = executor.execute(transfer_actions?);
+
+            Ok(Response::new()
+                .add_messages(transfer_msgs)
+                .add_attribute("total_amount", asset.amount.to_string())
+                .add_attribute("action", "charge_fixed_amount_penalty")
+                .add_attribute("remainder is", reaminder.to_string()))
+        }
+        Penalty::Daily {
+            asset: _,
+            split_between_friends: _,
+        } => {
+            // Not sure what the exact implementation should be here.
+            // Is it that for this variant we want to only charge_penalty at the end of the
+            // challenge? If so how do we determine when the challenge has come to an end?
+            //let _transfer_action = bank.transfer(vec![asset], &admin_address)?;
+            Ok(Response::new().add_attribute("action", "charge_daily_penalty"))
+        }
+    }
+}
+
+fn veto_vote(
+    deps: DepsMut,
+    info: MessageInfo,
+    _env: Env,
+    challenge_id: u64,
+    vote: Vote<String>,
+) -> AppResult {
+    if info.sender.to_string() != ADMIN.load(deps.storage)? {
```

Comment:
Do we have 2 admins? The primary and this secondary one?

Reply:
This was old code, the `ADMIN` state is gone now.